### PR TITLE
[PY] feat: Twenty Questions

### PR DESCRIPTION
**Pull request by BMS-geodev**
## Linked issues

closes: #316

## Details

A python implementation of the existing JS "Twenty Questions" sample bot. 
The bot leads the user through the twenty questions guessing game, with the user attempting to guess a secret word via successive guesses. Driven through the Teams toolkit, the sample launches a browser window where a instance of teams chat runs the bot locally. See screenshot for interaction.

#### Change details

Ported the code found in `/js/samples/03.ai-concepts/a.twentyQuestions/...` to Python, with usage of the existing Python implementation of things like`Application`, `ActionPlanner`, `context`, and so on. 

**code snippets**:
None.

**screenshots**:
image

## Attestation Checklist

- [X] My code follows the style guidelines of this project

- I have checked for/fixed spelling, linting, and other errors
- I have commented my code for clarity
- I have made corresponding changes to the documentation (updating the doc strings in the code is sufficient)
- My changes generate no new warnings
- I have added tests that validates my changes, and provides sufficient test coverage. I have tested with:
  - Local testing
  - E2E testing in Teams
- New and existing unit tests pass locally with my changes

### Additional information

> Feel free to add other relevant information below

### CODE REVIEW

**Review by lilyydu — COMMENTED** [review]
Awesome job setting this up! Couple of fixes- let me know if you want me to clarify anything :)

File: python/samples/04.ai.a.twentyQuestions/appPackage/manifest.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
+{
+    "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.15/MicrosoftTeams.schema.json",
+    "version": "1.0.0",
+    "manifestVersion": "1.15",
+    "id": "${{TEAMS_APP_ID}}",
+    "packageName": "com.package.name",
+    "name": {
+        "short": "LightBot-${{TEAMSFX_ENV}}",
```
[the remark this answers is not in the document]
Comment (BMS-geodev) [reply python/samples/04.ai.a.twentyQuestions/appPackage/manifest.json:8]: Done!

File: python/samples/04.ai.a.twentyQuestions/src/bot.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+config = Config()
+
+if config.OPENAI_KEY is None and config.AZURE_OPENAI_KEY is None:
+    raise RuntimeError(
+        "Missing environment variables - please check that OPENAI_KEY or AZURE_OPENAI_KEY is set."
+    )
+
+@dataclass
+class ConversationState:
```
[the remark this answers is not in the document]
Comment (BMS-geodev) [reply python/samples/04.ai.a.twentyQuestions/src/bot.py:37]: done

**Comment by BMS-geodev** [comment]
@lilyydu addressed all your feeback, thank you!

Followed your idea to lace print statements into teams/app.py, and have a finding where I got stumped 

here is the terminal feed:
```
turn_state_factory begin
======== Running on http://localhost:3978 ========
(Press CTRL+C to quit)
process begin
on_turn begin
_start_long_running_call begin
_on_turn begin
_on_turn start typing
_on_turn turn_state_factory
_on_activity begin
_on_activity []
_on_activity []
```
I hit the end of the thread in _on_activity()
```
    async def _on_activity(self, context: TurnContext, state: StateT) -> Tuple[bool, int]:
        matches = 0
        print("_on_activity begin")
        # ensure we handle invokes first
        routes = filter(lambda r: not r.is_invoke and r.selector(context), self._routes)
        print(f"_on_activity {list(routes)}")
        invoke_routes = filter(lambda r: r.is_invoke and r.selector(context), self._routes)
        print(f"_on_activity {list(invoke_routes)}")
        for route in invoke_routes:
            print("_on_activity invoke_routes")
            if route.selector(context):
                matches = matches + 1

                if not await route.handler(context, state):
                    return False, matches

        for route in routes:
            print("_on_activity routes")
            if route.selector(context):
                matches = matches + 1

                if not await route.handler(context, state):
                    return False, matches

        return True, matches
        ```
        specifically it seems like routes, and invoke_routes are not finding any result from the filter. 
        
        what do you think?
        
        I stopped logging to get your read, had to stop somewhere hah

File: .vscode/launch.json
Comment (lilyydu) [inline .vscode/launch.json:1]: think you accidentally committed this!

Comment (BMS-geodev) [reply .vscode/launch.json:1]: so weird, thanks

**Comment by BMS-geodev** [comment]
@microsoft-github-policy-service agree company="Awarity Inc."

**Review by lilyydu — APPROVED** [review]

**Pull request merged**